ufl/algorithms/map_integrands.py

```diff
@@ -19,20 +19,24 @@
 from ufl.constantvalue import Zero
 
 
-def map_integrands(function, form, only_integral_type=None):
+def map_integrands(function, form, only_integral_type=None,
+                   reconstruct_integral_domain=False):
     """Apply transform(expression) to each integrand
     expression in form, or to form if it is an Expr.
     """
     if isinstance(form, Form):
-        mapped_integrals = [map_integrands(function, itg, only_integral_type)
+        mapped_integrals = [map_integrands(function, itg, only_integral_type,
+                                           reconstruct_integral_domain)
                             for itg in form.integrals()]
         nonzero_integrals = [itg for itg in mapped_integrals
                              if not isinstance(itg.integrand(), Zero)]
         return Form(nonzero_integrals)
     elif isinstance(form, Integral):
         itg = form
         if (only_integral_type is None) or (itg.integral_type() in only_integral_type):
-            return itg.reconstruct(function(itg.integrand()))
+            integrand = function(itg.integrand())
+            domain = integrand.ufl_domain() if reconstruct_integral_domain else None
+            return itg.reconstruct(integrand, domain=domain)
```

Comment:
You need this because the domain may also have "heavy" data?

Reply:
Yep.

Reply:
This isn't a very appealing fix though because I now have this `reconstruct_integral_domain` keyword argument hanging around. I initially tried without but then `test/test_mixed_function_space.py::test_mixed_functionspace` was failing with `ufl.log.UFLException: Found domains with different geometric dimensions`.

Reply:
This seems like a bug though, because AIUI your change doesn't touch anything that should affect that test?

Reply:
That's possible. I'll investigate.

Reply:
So the issue here is that in `test_mixed_functionspace.py` we have forms defined over multiple domains (1D, 2D and 3D). This means that just calling `integrand.ufl_domain()` fails since there is more than one domain being considered. I have no real idea what to do about this. @wence- any thoughts?